### DISCUSSION TITLE
`render:*` directive

### DISCUSSION THREAD

**Discussion by boehs** [withastro/roadmap#126] [Proposal]
A render directive that allows using renderers inline. This would replace the markdown component and the `lang="scss"`/`lang="ts"` attributes, it would also consequently add support for inline react, vue, svelte, etc

## Examples

- `render:md`
- `render:scss`
- `render:ts`
- `render:vue`
- `render:react client:only`?

## Why is this needed

- `<Markdown>` component is very hacky and loosely speaking, an objectively bad way to do things 
- Inline svelte/react has been a requested feature
- `lang="ts"` is not the astro way of doing things

```html
<p render:md>
# h1
</p>
```

**Comment by Princesseuh** [comment]
Personally, I'm +1 on everything (even though I will dearly miss lang="ts" coming from Vue where it works so nicely) except for inline renderers

I get that it would be fun and perhaps convenient but I think it would make for really messy files and projects in addition of it being practically impossible to provide proper editor support. I see more downsides than upsides (which I actually kinda don't see any outside of the fun factor)

**Comment by FredKSchott** [comment]
## Example

```
<article render:md>
  # Hello, World!
</article>
<Fragment render:svelte>
  <script>...
  <style>...
  <h1>Hello, World</h1>
</Fragment>
```

## Notes

- `is:raw` - allows you to put anything you'd like into a component
- implementation - could pull into another virtual file internally?
- naming - `render:X` vs. `use:render={X}` or `is:lang={X}` or `has:...`
- relevant - we currently support `<style lang="scss">`, this is a similar feature. Could become `<style is:lang={X}>` in this proposal (naming tbd)
- idea - could we refactor our Sass support to be more pluggable and create a system internally that people could hook into, so that they could experiment more with this?
- idea - Markdown component is currently ~hacky~ not loved by all, could this be a way to finalize our inline markdown support for v1.0?

**Comment by ghost** [comment]
I think that `render:react` example deserves some exploration. For instance, how would a rendered React function operate? And how would you pass props to it?

Let's take a simple counter as an example:

```jsx
function LabelledCounter({ buttonLabel }) {
  const [count, setCount] = useState(0);

  return (
    <div>
      <p>You clicked {count} times</p>
      <button onClick={() => setCount(count + 1)}>
        {buttonLabel}
      </button>
    </div>
  );
}
```

How could I "call" this component using `render` tags? I suppose we could pass the function as an argument like so:

```astro
<Astro.fragment render:react={function LabelledCounter(...) {...})} buttonLabel="Count up" client:load />
```

... but this might be leaving the domain of what renderers "should" be. You'll also notice I've passed a component prop (`buttonLabel`) directly on the `Astro.fragment` for that `render:react` directive to pick up. As far as I understand, Astro directives can't pick up on miscellaneous props like this. And if they did, it wouldn't be easy to differentiate props for the parent element vs. props for the child. Here's an example:

```astro
<div render:react={function LabelledCounter(...) {...})} class="stylish" aria-label="neat" client:load />
```

☝️ Are `class` and `aria-label` attributes on my `div`? Are they _also_ props for the component I'm rendering? It's a bit confusing when parent and child are merged like this 😕 

## Question: should render be a directive, or a new Astro component?
Lets re-iterate that question from earlier: how can a `render` directive receive arguments, like a component, component props, or a `client:load` directive? It seems `render` starts to leave the world of directives entirely and become more... component-like.

I'm _not_ suggesting this is the final API, but I want to propose the idea: a new `Astro.render` component. This is loosely inspired by `Astro.fragment`, and offers a more generic alternative to components like `Markdown`, `Icon`, `Image`, etc. I'll also call out [11ty's `renderTemplate` shortcode](https://www.11ty.dev/docs/plugins/render/) as inspiration here.

Here's a few use cases that come to mind:

```astro
<!--An "Icon" component-->
<Astro.render as:svg local:src="/path/to/cool.svg" />
```

```astro
<!--Write markdown inline-->
<Astro.render as:markdown>
# A markdown component!
</Astro.render>

<!--Or render a file-->
<Astro.render as:markdown local:src="/path/to/local.md" />
```

```astro
<!--Inline React, Preact, Solid, etc-->
<Astro.render as:react={function LabelledCounter(...) {...})} buttonLabel="Count up" client:load />
```

```astro
<!--Optimized image: take in a source image, and generate a "picture" element with multiple image formats and sizes-->
<Astro.render as:picture local:src="/path/to/heckin-chonker.png" widths={[600, 800, 1000]} />
```

Using a component accomplishes a few things:
- Simplifies `Astro.fragment` + a render directive to a single component
- Makes prop passing a bit more obvious. Instead of trying to pass numerous props to a directive, we can apply props to a component

...And leaves some more unanswered questions:
- How will intellisense know which props to expect?
- How would we handle unnamed props? For example, knowing to use `as:react={Component}` may be confusing to new users.
- Isn't this even more like a shortcode? (the answer is yes. And that should be a pro 😉)

I also want to mention: I 100% support this RFC! Even limiting `render` to a directive, it would hopefully prevent a proliferation of components by file type (`Markdown`, `Icon`, `Image`, etc), and let you use your favorite templating language in-line without much trouble ❤️

**Reply by Enteleform** [reply]
I really dig the versatility of this approach.

What do you think about having `as` defaults/fallbacks when `src` is present, based on the file extension?

For example:
```html
<Astro.render local:src="/path/to/cool.svg"/>
is equivalent to
<Astro.render as:svg local:src="/path/to/cool.svg"/>
```

But `as` is still available for cases where the defaults might not work:
```html
<Astro.render as:markdown local:src="/path/to/article.w͇̋̀e͑ǐr̶̟̋d̉̕-̗ͩ̕c͑u͑͢s̸ͤţ̖̐ő̸͉m-̻͒̕ḛ̌͜x̋͏tͬ̀e̡ͦn̛͛s͗̕i̷̠̅ó̶̪n"/>
```
